kolibri/core/auth/tasks.py

```diff
@@ -1,118 +1,168 @@
-        _, filepath = mkstemp(dir=temp_dir, suffix=".upload")
-        with open(filepath, "w+b") as dest:
-            filepath = dest.name
-            for chunk in upload.file.chunks():
-                dest.write(chunk)
-    else:
-        tmpfile = upload.file.temporary_file_path()
-        filename = ntpath.basename(tmpfile)
-        filepath = os.path.join(temp_dir, filename)
-        shutil.copy(tmpfile, filepath)
-    return filepath
-
-
-def validate_importusersfromcsv(request, request_data):
-    temp_dir = os.path.join(conf.KOLIBRI_HOME, "temp")
-    if not os.path.isdir(temp_dir):
-        os.mkdir(temp_dir)
-
-    locale = get_language_from_request(request)
-    # the request must contain either an object file
-    # or the filename of the csv stored in Kolibri temp folder
-    # Validation will provide the file object, while
-    # Importing will provide the filename, previously validated
-    if not request.FILES:
-        filename = request_data.get("csvfile", None)
-        if filename:
-            filepath = os.path.join(temp_dir, filename)
-        else:
-            return HttpResponseBadRequest("The request must contain a file object")
-    else:
-        if "csvfile" not in request.FILES:
-            return HttpResponseBadRequest("Wrong file object")
-        filepath = manage_fileobject(request, temp_dir)
-
-    userid = request.user.pk
-    facility_id = request_data.get("facility_id", None)
-
-    job_metadata = {
-        "type": "IMPORTUSERSFROMCSV",
-        "started_by": userid,
-        "facility": facility_id,
-    }
-
-    job_args = []
-    if request_data.get("dryrun") is not None:
-        job_args.append("--dryrun")
-    if request_data.get("delete") is not None:
-        job_args.append("--delete")
-    job_args.append(filepath)
-
-    return {
-        "job_args": job_args,
-        "facility": facility_id,
-        "userid": userid,
-        "locale": locale,
-        "extra_metadata": job_metadata,
-    }
+    def validate(self, data):
+        if data.get("csvfile") and data.get("csvfilename"):
+            raise serializers.ValidationError(
+                "Only one of csvfile or csvfilename can be specified"
+            )
+        if not data.get("csvfile") and not data.get("csvfilename"):
+            raise serializers.ValidationError(
+                "One of csvfile or csvfilename must be specified"
+            )
+        filepath = data.get("csvfile") or data.get("csvfilename")
+        args = [filepath]
+        if data.get("dryrun"):
+            args.append("--dryrun")
+        if data.get("delete"):
+            args.append("--delete")
+        kwargs = {"locale": data.get("locale"), "facility": data.get("facility_id")}
+        if "user" in self.context:
+            kwargs["userid"] = self.context["user"].id
+        return {
+            "args": args,
+            "kwargs": kwargs,
+            "facility_id": data.get("facility_id"),
+        }
```

Comment:
In theory, the validator can return more `args`, `kwargs` than mentioned in the Serializer, right? So how exactly Serialier is serving the purpose of introspecting of expected arguments for a task function?

Reply:
The serializer/validator is an interface between the API request data and the task function, so it allows introspection of what is required to be sent to the API endpoint in order to initialize a task.

As tasks can still be directly enqueued in the backend without using the API, the function signature of the task function itself is still available for that purpose, although it was convenient in at least one point in this refactor to use the validation functionality for a task enqueued entirely from the backend as well.

Reply:
I am skeptical regarding the use of serializers to just introduce introspection abilities. IMO like any other API, documentation should be a single source of truth for conveying what's expected and what will be the response in various scenarios.

Reply:
Moreover the task's function signature is enough to know for our frontend devs to understand what's required for the task to be enqueued. That's what function signature's purpose is to convey required data. And regarding their type, we should use code documentation. 

I feel serializers have introduced more complexity than there should be...

Reply:
> the task's function signature is enough to know for our frontend devs about what's required for the task to be enqueued

This was never the case though - the validator functions that were already being used did not just validate and pass through frontend arguments to the task function. They validated, added additional data, and did so with no explicit detailing of what arguments they expect.

By turning from validator function to validator serializers, we now have an explicit delineation of what is expected. This can be inspected by looking at the code, or by looking at `/api_explorer` and examining the task endpoint, which now gives a full spec of the accepted task types and their arguments. This also means that frontend devs do not need to delve through Python code nor understand it to have a clear spec of the task arguments.

Reply:
@rtibbles makes sense!